internal/logsink/logsink.go

```diff
@@ -0,0 +1,288 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package logsink
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/juju/loggo/v2"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/logger"
+)
+
+const (
+	stateFlushed = "flushed"
+	stateTicked  = "ticked"
+)
+
+// NewWriterFunc is a function that creates a new writer.
+type NewWriterFunc func() (io.WriteCloser, error)
+
+// LogSink is a loggo.Writer that writes log messages to a file.
+type LogSink struct {
+	tomb           tomb.Tomb
+	internalStates chan string
+
+	writer io.Writer
+
+	batchSize     int
+	flushInterval time.Duration
+
+	inLogEntry   chan loggo.Entry
+	inLogRecords chan []logger.LogRecord
+
+	out chan []logRecord
+}
+
+// NewLogSink creates a new log sink that writes log messages to a file. There
+// can only be one writer writing to the same file at a time, otherwise bytes
+// will be written to the file in an interleaved manner (junk data). LogSink
+// writer will write log messages as JSON objects, one per line, even if the log
+// message is multiline. The batchSize parameter specifies the minimum number of
+// log messages to batch before writing to the underlying writer. The number of
+// entires can far exceed the batchSize if the log messages are large.
+func NewLogSink(writer io.Writer, batchSize int, flushInterval time.Duration) *LogSink {
+	return newLogSink(writer, batchSize, flushInterval, nil)
```

Comment:
We take a `io.Writer` here and because this is a worker, it is taking ownership. Therefore, this should be a `io.WriterCloser` and be `Close()`d when the worker exits.

Reply:
Fair. I'll update the description outlining ownership as well.